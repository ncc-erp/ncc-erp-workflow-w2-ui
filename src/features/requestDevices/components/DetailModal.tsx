import {
  Divider,
  HStack,
  Heading,
  Image,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Spinner,
  Text,
  Button,
} from '@chakra-ui/react';
import Logo from 'assets/images/ncc_logo.png';
import styles from './style.module.scss';
import { Request } from 'models/request';
import { useGetRequestDetail } from 'api/apiHooks/requestHooks';
import { useMemo, useState } from 'react';
import { RequestInput } from 'features/Tasks/components/RequestInput';
import { TextGroup } from 'common/components/TextGroup/TextGroup';
import { isObjectEmpty, formatDate, getColorByStatus } from 'utils';
import { WorkflowModal } from 'common/components/WorkflowModal';
import { RequestStatus } from 'common/enums';

interface IDetailModalProps {
  isOpen: boolean;
  onClose: () => void;
  requestDetail: Request;
}

export const RequestDetailModal = ({
  isOpen,
  onClose,
  requestDetail,
}: IDetailModalProps) => {
  const { data, isLoading } = useGetRequestDetail(requestDetail.id);

  const [requestWorkflow, setRequestWorkflow] = useState<string>('');
  const [isOpenWorkflow, setOpenWorkflow] = useState(false);

  const onActionViewWorkflow = (workflowId: string) => () => {
    setRequestWorkflow(workflowId);
    setOpenWorkflow(true);
  };

  const { inputRequestDetail, inputRequestUser, tasks } = useMemo(() => {
    const { typeRequest, input, tasks, workInstanceId } = data || {};
    const { RequestUser, Request } = input || {};

    return {
      tasks,
      inputRequestUser: RequestUser,
      inputRequestDetail: Request,
      typeRequest,
      workInstanceId,
    };
  }, [data]);

  const rejectReason = useMemo(() => {
    if (!tasks || requestDetail?.status !== RequestStatus.Rejected) {
      return null;
    }

    return tasks?.find((task) => task.reason)?.reason;
  }, [requestDetail?.status, tasks]);

  const hasInputRequestData: boolean = useMemo(() => {
    return !isObjectEmpty(inputRequestDetail);
  }, [inputRequestDetail]);

  if (isLoading) {
    return (
      <Modal isOpen={isOpen} onClose={onClose}>
        <ModalOverlay />
        <ModalContent p="10px" maxW="700px">
          <ModalBody>
            <div className={styles.containerSpinner}>
              <Spinner color="red.500" size="xl" />
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>
    );
  }

  return (
    <>
      <Modal isOpen={isOpen} onClose={onClose} blockScrollOnMount={false}>
        <ModalOverlay />
        <ModalContent p="10px" maxW="700px">
          <ModalHeader>
            <HStack>
              <Image h="45px" src={Logo} />
              <Heading ml={1} w="550px">
                <Text color="primary" fontSize={18}>
                  {requestDetail?.workflowDefinitionDisplayName}
                </Text>
                <Text fontSize={16} fontWeight={400} mt={1.5}>
                  Request Details
                </Text>
              </Heading>
            </HStack>
            <Button mt={2} onClick={onActionViewWorkflow(requestDetail.id)}>
              View Workflow Detail
            </Button>
          </ModalHeader>
          <ModalCloseButton mt="15px" mr="10px" />

          <ModalBody className={styles.modalBody} pr={2}>
            <Divider mt={2} mb={3} />
            <div className={styles.container}>
              <Text
                mb="10px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Request input
              </Text>
              <div className={styles.wrapper}>
                {hasInputRequestData && inputRequestDetail && (
                  <RequestInput inputRequestDetail={inputRequestDetail} />
                )}
              </div>

              <Divider mt={2} mb={3} />
              <Text
                mb="10px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Request user
              </Text>
              <div className={styles.wrapper}>
                {inputRequestUser ? (
                  <>
                    <TextGroup label="Name" content={inputRequestUser?.name} />
                    <TextGroup
                      label="Email"
                      content={inputRequestUser?.email}
                    />
                    <TextGroup
                      label="Branch name"
                      content={inputRequestUser?.branchName}
                    />
                  </>
                ) : (
                  <Text fontSize="15px">Not found!</Text>
                )}
              </div>

              <Divider mt={2} mb={3} />
              <Text
                mb="15px"
                fontWeight={600}
                fontStyle="italic"
                color="primary"
              >
                Detail
              </Text>

              <div className={styles.wrapper}>
                <TextGroup
                  label="Request template"
                  content={requestDetail?.workflowDefinitionDisplayName}
                />
                <TextGroup
                  label="Status"
                  content={getColorByStatus(requestDetail?.status).status}
                  color={getColorByStatus(requestDetail?.status).color}
                />

                {requestDetail?.currentStates.length > 0 && (
                  <TextGroup
                    label="Current state"
                    content={requestDetail?.currentStates.join(', ')}
                  />
                )}
                {requestDetail?.stakeHolders.length > 0 && (
                  <TextGroup
                    label="Stakeholders"
                    content={requestDetail?.stakeHolders.join(', ')}
                  />
                )}
                <TextGroup
                  label="Creation time"
                  content={
                    requestDetail?.createdAt
                      ? formatDate(new Date(requestDetail?.createdAt))
                      : ''
                  }
                />

                {rejectReason && (
                  <TextGroup label="Reason" content={rejectReason} />
                )}
              </div>
            </div>
          </ModalBody>
        </ModalContent>
      </Modal>
      {requestWorkflow && (
        <WorkflowModal
          isOpen={isOpenWorkflow}
          onClose={() => setOpenWorkflow(false)}
          workflow={`CompOnly?id=${requestWorkflow}`}
        />
      )}
    </>
  );
};
