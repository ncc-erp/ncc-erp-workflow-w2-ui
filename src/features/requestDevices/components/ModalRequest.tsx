import {
  ModalCloseButton,
  ModalContent,
  ModalOverlay,
  FormControl,
  ModalFooter,
  ModalHeader,
  FormLabel,
  ModalBody,
  Button,
  Select,
  Modal,
  Input,
} from "@chakra-ui/react";
import { formatsQuill, modulesQuill } from "common/constants/others";
import { InputTemplatesTypes } from "common/constants/type";
import { FormField, IFormRequest, IPropertyDefinition } from "common/interface";
import { Controller, useForm } from "react-hook-form";
import "react-quill/dist/quill.snow.css";
import ReactQuill from "react-quill";
import { useEffect } from "react";

interface Props {
  isOpen: boolean;
  onClose: () => void;
  requestModal: {
    title: string;
    inputs: IPropertyDefinition[];
  };
}

const ModalRequest = ({ isOpen, onClose, requestModal }: Props) => {
  const {
    handleSubmit,
    register,
    control,
    reset,
    formState: { isSubmitting },
  } = useForm();

  function onSubmit(values: Partial<IFormRequest>) {
    console.log(values);
  }

  useEffect(() => {
    reset();
  }, [isOpen, reset]);

  return (
    <Modal isOpen={isOpen} onClose={onClose} size={"xl"}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader pb={0}>{requestModal?.title}</ModalHeader>
        <ModalCloseButton />
        <form onSubmit={handleSubmit(onSubmit)}>
          <ModalBody pb={6}>
            {requestModal?.inputs?.map((item: IPropertyDefinition, index) => {
              const name = item?.name as FormField;
              return (
                <FormControl
                  key={index}
                  isRequired={item?.isRequired}
                  mt={4}
                  h={item?.type === InputTemplatesTypes.RICH_TEXT ? 275 : ""}
                >
                  <FormLabel>
                    {item?.name?.replace(/([A-Z])/g, " $1").trim()}
                  </FormLabel>
                  {item?.type === InputTemplatesTypes.OFFICE_LIST && (
                    <Controller
                      control={control}
                      name={name}
                      render={({ field }) => (
                        <Select placeholder="Select office" {...field}>
                          <option value="office1">Office 1</option>
                          <option value="office2">Office 2</option>
                          <option value="office3">Office 3</option>
                        </Select>
                      )}
                    />
                  )}
                  {item?.type === InputTemplatesTypes.MY_PROJECT && (
                    <Controller
                      control={control}
                      name={name}
                      render={({ field }) => (
                        <Select placeholder="Select project" {...field}>
                          <option value="project1">Project 1</option>
                          <option value="project2">Project 2</option>
                          <option value="project3">Project 3</option>
                        </Select>
                      )}
                    />
                  )}
                  {item?.type === InputTemplatesTypes.TEXT && (
                    <Input type="text" {...register(name)} />
                  )}
                  {item?.type === InputTemplatesTypes.RICH_TEXT && (
                    <Controller
                      control={control}
                      name={name}
                      render={({ field }) => (
                        <ReactQuill
                          formats={formatsQuill}
                          modules={modulesQuill}
                          style={{ height: "200px" }}
                          theme="snow"
                          {...field}
                        />
                      )}
                    />
                  )}
                  {item?.type === InputTemplatesTypes.DATE_TIME && (
                    <Input
                      placeholder="Select Date"
                      type="date"
                      {...register(name)}
                    />
                  )}
                </FormControl>
              );
            })}
          </ModalBody>
          <ModalFooter>
            <Button onClick={onClose} mr={3}>
              Cancel
            </Button>
            <Button colorScheme="blue" isLoading={isSubmitting} type="submit">
              Save
            </Button>
          </ModalFooter>
        </form>
      </ModalContent>
    </Modal>
  );
};

export default ModalRequest;
