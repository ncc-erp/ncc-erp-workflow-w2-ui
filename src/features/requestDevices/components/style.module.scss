.container {
    display: flex;
    flex-direction: column;
    margin-top: 10px;
}

.wrapper{
    display: flex;
    width: 100%;
    flex-direction: row;
    flex-wrap: wrap;
}

.wrapper  > div {
    width: 50%;
    margin-bottom: 10px !important;
}

.modalBody {
    max-height: 70vh;
    overflow: scroll;
}

.listItem {
    white-space: nowrap;      
    overflow: hidden;         
    text-overflow: ellipsis;  
    width: 100%;

    display: flex;
    align-items: center;
}

.actions {
    display: flex;
    align-items: center;
    flex-direction: row;
    width: 100%;
    flex-wrap: wrap;
    gap: 5px;
    margin-top: 15px;

    .spinner {
        min-width: 24px;
        margin-right: 5px;
    }
}

.containerSpinner {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: 10px;
    min-width: 30vw;
    min-height: 50vh;
}