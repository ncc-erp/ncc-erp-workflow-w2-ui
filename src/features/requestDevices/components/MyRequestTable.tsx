import {
  ColumnDef,
  SortingState,
  createColumnHelper,
} from '@tanstack/react-table';
import { useQueryClient } from '@tanstack/react-query';
import {
  Box,
  Button,
  Center,
  HStack,
  Spacer,
  Spinner,
  Wrap,
  WrapItem,
} from '@chakra-ui/react';
import {
  useCancelRequest,
  useDeleteRequest,
  useMyRequests,
  useRequestTemplates,
} from 'api/apiHooks/requestHooks';
import { SelectField } from 'common/components/SelectField';
import { Table } from 'common/components/Table/Table';
import { RequestSortField, RequestStatus, SortDirection } from 'common/enums';
import { FilterRequestParams, Request } from 'models/request';
import { useEffect, useMemo, useState } from 'react';
import { Pagination } from 'common/components/Pagination';
import { QueryKeys, noOfRows } from 'common/constants';
import { PageSize } from 'common/components/Table/PageSize';
import { ShowingItemText } from 'common/components/Table/ShowingItemText';
import { RowAction } from 'features/requestDevices/components/RowAction';
import { EmptyWrapper } from 'common/components/EmptyWrapper';
import { useRecoilValue } from 'recoil';
import { appConfigState } from 'stores/appConfig';
import { toast } from 'common/components/StandaloneToast';
import { ModalConfirm } from 'common/components/ModalConfirm';
import { useCurrentUser } from 'hooks/useCurrentUser';
import { formatDate } from 'utils';
import { useIsAdmin } from 'hooks/useIsAdmin';
import { RequestDetailModal } from './DetailModal';

const initialSorting: SortingState = [
  {
    id: RequestSortField.createdAt,
    desc: true,
  },
];

export const MyRequestTable = () => {
  const currentUser = useCurrentUser();
  const initialFilter: FilterRequestParams = {
    Status: '',
    WorkflowDefinitionId: '',
    sorting: [RequestSortField.createdAt, 'desc'].join(' '),
    skipCount: 0,
    maxResultCount: +noOfRows[0].value,
    RequestUser: currentUser?.sub[0],
    StakeHolder: '',
  };
  const { sideBarWidth } = useRecoilValue(appConfigState);
  const [filter, setFilter] = useState<FilterRequestParams>(initialFilter);
  const [sorting, setSorting] = useState<SortingState>(initialSorting);
  const { data, isLoading } = useMyRequests(filter);
  const { data: requestTemplateData } = useRequestTemplates();
  const { items: requests = [], totalCount = 0 } = data ?? {};
  const { items: requestTemplates = [] } = requestTemplateData ?? {};
  const { skipCount, maxResultCount } = filter;
  const currentPage = (maxResultCount + skipCount) / maxResultCount;
  const columnHelper = createColumnHelper<Request>();
  const isAdmin = useIsAdmin();

  const queryClient = useQueryClient();
  const deleteRequestMutation = useDeleteRequest();
  const cancelRequestMutation = useCancelRequest();
  const [isOpen, setIsOpen] = useState(false);
  const [isOpenDetails, setOpenDetails] = useState(false);
  const [modalTitle, setModalTitle] = useState('');
  const [modalDescription, setModalDescription] = useState('');
  const [actionType, setActionType] = useState('');
  const [requestId, setRequestId] = useState('');
  const [requestDetails, setRequestDetails] = useState<Request>();

  const statusOptions = useMemo(() => {
    const defaultOptions = {
      value: '',
      label: 'All status',
    };

    const options = Object.values(RequestStatus).map((value) => ({
      value,
      label: value == RequestStatus.Pending ? 'Pending' : value,
    }));

    return [defaultOptions, ...options];
  }, []);

  const requestTemplateOtions = useMemo(() => {
    const defaultOptions = {
      value: '',
      label: 'All types',
    };

    const options = requestTemplates.map(({ definitionId, displayName }) => ({
      value: definitionId,
      label: displayName,
    }));

    return [defaultOptions, ...options];
  }, [requestTemplates]);

  const myRequestColumns = useMemo(
    () =>
      [
        columnHelper.accessor('workflowDefinitionDisplayName', {
          id: 'workflowDefinitionDisplayName',
          header: () => <Box>Request template</Box>,
          enableSorting: false,
          cell: (info) => <Box>{info.getValue()}</Box>,
        }),
        columnHelper.accessor('userRequestName', {
          id: 'userRequestName',
          header: () => <Box>Request user</Box>,
          enableSorting: false,
          cell: (info) => info.getValue(),
        }),
        columnHelper.accessor('currentStates', {
          id: 'currentStates',
          header: () => <Box textAlign="center">Current states</Box>,
          enableSorting: false,
          cell: (info) => info.getValue().join('\n'),
        }),
        columnHelper.accessor('stakeHolders', {
          id: 'stakeHolders',
          header: 'Stake holders',
          enableSorting: false,
          cell: (info) => info.getValue().join(', '),
        }),
        columnHelper.accessor('createdAt', {
          id: 'createdAt',
          header: 'Created at',
          cell: (info) => formatDate(new Date(info.getValue())),
          sortDescFirst: true,
        }),
        columnHelper.accessor('lastExecutedAt', {
          id: 'lastExecutedAt',
          header: 'Last executed at',
          cell: (info) => formatDate(new Date(info.getValue())),
          sortDescFirst: true,
        }),
        columnHelper.accessor('status', {
          id: 'status',
          header: 'Status',
          enableSorting: false,
          cell: (info) => info.getValue(),
        }),
        columnHelper.display({
          id: 'actions',
          enableSorting: false,
          header: () => <Center w="full">Actions</Center>,
          cell: (info) => (
            <Center>
              <RowAction
                onViewDetails={onActionViewDetails(info.row.original)}
                onCancel={onAction(info.row.original.id, 'cancel')}
                onDelete={onAction(info.row.original.id, 'delete')}
              />
            </Center>
          ),
        }),
      ] as ColumnDef<Request>[],
    [columnHelper]
  );

  useEffect(() => {
    const { id, desc } = sorting?.[0] ?? {};
    const sort = `${id} ${desc ? SortDirection.desc : SortDirection.asc}`;

    setFilter((filter) => ({
      ...filter,
      sorting: sort,
      skipCount: 0,
    }));
  }, [sorting]);

  const onPageChange = (page: number) => {
    setFilter((filter) => ({
      ...filter,
      skipCount: filter.maxResultCount * (page - 1),
    }));
  };

  const onPageSizeChange = (pageSize: number) => {
    setFilter((filter) => ({
      ...filter,
      maxResultCount: pageSize,
      skipCount: 0,
    }));
  };

  const onTemplateStatusChange = (
    key: 'Status' | 'WorkflowDefinitionId' | 'RequestUser' | 'StakeHolder',
    value?: string
  ) => {
    setFilter({ ...filter, [key]: value, skipCount: 0 });
  };

  const onActionViewDetails = (request: Request) => () => {
    setRequestDetails(request);
    setOpenDetails(true);
  };

  const onAction = (requestId: string, type: 'delete' | 'cancel') => () => {
    setRequestId(requestId);
    setActionType(type);
    setModalTitle(`Confirm ${type} request`);
    setModalDescription(`Request will be ${type}ed. Do you confirm that?`);
    setIsOpen(true);
  };

  const handleConfirmation = async () => {
    setIsOpen(false);
    if (requestId.length === 0) return;

    const mutation =
      actionType === 'delete' ? deleteRequestMutation : cancelRequestMutation;
    const successMessage =
      actionType === 'delete'
        ? 'Deleted successfully!'
        : 'Cancelled successfully!';
    const errorMessage =
      actionType === 'delete' ? 'Delete failed!' : 'Cancel failed!';

    try {
      await mutation.mutateAsync(requestId);
      queryClient.invalidateQueries({ queryKey: [QueryKeys.FILTER_REQUEST] });
      toast({ title: successMessage, status: 'success' });
    } catch (error) {
      toast({ title: errorMessage, status: 'error' });
    }
  };

  return (
    <>
      <Box>
        <HStack
          w="full"
          pl="24px"
          pb="8px"
          alignItems="flex-end"
          flexWrap="wrap"
        >
          <Box w="220px">
            <SelectField
              size="sm"
              rounded="md"
              onChange={(e) =>
                onTemplateStatusChange('WorkflowDefinitionId', e.target.value)
              }
              options={requestTemplateOtions}
            />
          </Box>
          <Box w="112px">
            <SelectField
              size="sm"
              rounded="md"
              onChange={(e) => onTemplateStatusChange('Status', e.target.value)}
              options={statusOptions}
            />
          </Box>
        </HStack>
        {isAdmin && (
          <Wrap pl="24px" pt="8px">
            <WrapItem>
              <Button
                size={'md'}
                colorScheme={filter.RequestUser ? 'green' : 'gray'}
                onClick={() => {
                  if (filter.RequestUser) {
                    setFilter({ ...filter, RequestUser: '' });
                  } else {
                    setFilter({ ...filter, RequestUser: currentUser?.sub[0] });
                  }
                }}
                fontSize="sm"
                fontWeight="medium"
                mr={2}
              >
                Only my request
              </Button>
            </WrapItem>
          </Wrap>
        )}
        {isLoading ? (
          <Center h="200px">
            <Spinner mx="auto" speed="0.65s" thickness="3px" size="xl" />
          </Center>
        ) : (
          <EmptyWrapper
            isEmpty={!requests.length}
            h="200px"
            fontSize="xs"
            message={'No requests found!'}
          >
            <Box
              p="20px 30px 0px 24px"
              overflowX="auto"
              w={{ base: `calc(100vw - ${sideBarWidth}px)`, lg: 'auto' }}
            >
              <Table
                columns={myRequestColumns}
                data={requests}
                sorting={sorting}
                onSortingChange={setSorting}
              />
            </Box>
          </EmptyWrapper>
        )}
        <HStack
          p="20px 30px 20px 30px"
          justifyContent="space-between"
          flexWrap="wrap"
        >
          <HStack alignItems="center" spacing="6px" flexWrap="wrap">
            <PageSize noOfRows={noOfRows} onChange={onPageSizeChange} />
            <Spacer w="12px" />
            <ShowingItemText
              skipCount={filter.skipCount}
              maxResultCount={filter.maxResultCount}
              totalCount={totalCount}
            />
          </HStack>
          <Pagination
            total={totalCount}
            pageSize={filter.maxResultCount}
            current={currentPage}
            onChange={onPageChange}
            hideOnSinglePage
          />
        </HStack>
      </Box>
      <ModalConfirm
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        onConfirm={handleConfirmation}
        title={modalTitle}
        description={modalDescription}
      />
      {requestDetails && (
        <RequestDetailModal
          isOpen={isOpenDetails}
          onClose={() => setOpenDetails(false)}
          requestDetail={requestDetails}
        />
      )}
    </>
  );
};
