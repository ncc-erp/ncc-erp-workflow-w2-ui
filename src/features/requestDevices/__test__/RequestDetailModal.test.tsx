import { render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RequestDetailModal } from '../components/DetailModal';
import { RequestStatus } from 'common/enums';

jest.mock('../../../api/apiHooks/index', () => ({
  useAxios: jest.fn(),
}));

jest.mock('../../../api/axiosInstant', () => ({
  VITE_API_BASE_URL: '/api',
}));

jest.mock('common/components/WorkflowModal', () => ({
  VITE_PROXY_SERVER_URL: 'http://localhost:4433',
}));

jest.mock('api/apiHooks/requestHooks', () => ({
  useGetRequestDetail: jest.fn().mockReturnValue({
    isLoading: false,
    data: {
      typeRequest: 'Office Equipment Request',
      input: {
        Request: {
          CurrentOffice: 'ĐN',
          Equipment: 'test current state',
          Reason: 'test',
        },
        EmailTemplate:
          '\n<style>\n    p {\n        margin: 0;\n    }\n</style>\n<div style="margin: 13px 0">\n    <b>Văn phòng đang làm việc</b>: Đà Nẵng\n</div>\n<div style="margin: 13px 0">\n    <b>Thiết bị cần request</b>: test current state\n</div>\n<div style="display: flex; margin: 13px 0;">\n    <b>Lý do:</b>&#160;test\n</div>\n',
        RequestUser: {
          email: 'nhan.huynhba@ncc.asia',
          name: 'Nhan Huynh Ba',
          project: null,
          pm: 'thien.dang@ncc.asia',
          headOfOfficeEmail: 'thien.dang@ncc.asia',
          projectCode: 'asm',
          branchName: 'Đà Nẵng',
          branchCode: 'ĐN',
          id: '3a0d3270-71a8-69d7-265b-ed35f5cc9960',
        },
      },
      workInstanceId: '3a0e24f1-b739-1627-f78a-cb2bdd068279',
      tasks: [
        {
          id: '3a0e24f1-b9a4-732f-65f2-836b6d9b6af9',
          workflowInstanceId: '3a0e24f1-b739-1627-f78a-cb2bdd068279',
          workflowDefinitionId: '3a0b89d4-93b5-232f-964d-ffa129064cc6',
          email: null,
          status: 0,
          name: 'Office Equipment Request',
          description: 'Branch Manager Makes Decision',
          dynamicActionData: null,
          reason: null,
          creationTime: '2023-10-09T04:49:34.630211Z',
          otherActionSignals: [],
          emailTo: null,
          author: '3a0d3270-71a8-69d7-265b-ed35f5cc9960',
          authorName: null,
        },
      ],
    },
  }),
  useOffices: jest.fn().mockReturnValue({
    data: [
      {
        displayName: 'Đà Nẵng',
        code: 'ĐN',
        headOfOfficeEmail: 'thien.dang@ncc.asia',
      },
    ],
  }),
  useUserList: jest.fn().mockReturnValue({
    data: [
      {
        name: 'Ân Bùi Hoàng',
        email: 'an.buihoang@ncc.asia',
      },
    ],
  }),
}));

jest.mock('api/apiHooks/index', () => ({
  useDownloadFile: jest.fn(() => jest.fn()),
}));

const requestDetails = {
  id: '3a0e24f1-b739-1627-f78a-cb2bdd068279',
  workflowDefinitionId: '3a0b89d4-93b5-232f-964d-ffa129064cc6',
  workflowDefinitionDisplayName: 'Office Equipment Request',
  userRequestName: 'Nhan Huynh Ba',
  createdAt: '2023-10-09T04:49:34.009027Z',
  lastExecutedAt: '2023-10-09T04:49:38.214996Z',
  status: RequestStatus.Approved,
  stakeHolders: ['Nhan Huynh Ba'],
  currentStates: ['Branch Manager makes decision'],
};

describe('My Request Detail', () => {
  const queryClient: QueryClient = new QueryClient();

  test('should match snapshot when rendering', async () => {
    const { baseElement } = render(
      <QueryClientProvider client={queryClient}>
        <RequestDetailModal
          isOpen={true}
          onClose={() => jest.fn()}
          requestDetail={requestDetails}
        />
      </QueryClientProvider>
    );
    expect(baseElement).toMatchSnapshot();
  });

  describe('My Request Detail with requestDetails', () => {
    beforeEach(() => {
      render(
        <QueryClientProvider client={queryClient}>
          <RequestDetailModal
            isOpen={true}
            onClose={() => jest.fn()}
            requestDetail={requestDetails}
          />
        </QueryClientProvider>
      );
    });

    it('Should have a Title when Request Detail loaded.', () => {
      expect(screen.getAllByText(/Office Equipment Request/i)).toBeTruthy();
    });

    describe('Request input attributes', () => {
      it('Should have a Request input attributes when Request Detail loaded.', async () => {
        screen.findByText(/Request input/i);
      });

      it('Should have a Current Office label in Request input', () => {
        screen.findByLabelText(/Current Office/i);
      });

      it('Should have a Reason label in Request input', () => {
        screen.findByLabelText(/Reason/i);
      });

      it('Should have a Equipment label in Request input', () => {
        screen.findByLabelText(/Equipment/i);
      });
    });

    describe('Request user attributes', () => {
      it('Should have a Request user attributes when Request Detail loaded.', async () => {
        screen.findByText(/Request user/i);
      });

      it('Should have a Name label in Request user', () => {
        screen.findByLabelText(/Name/i);
      });

      it('Should have a Email label in Request user', () => {
        screen.findByLabelText(/Email/i);
      });

      it('Should have a Branch name label in Request user', () => {
        screen.findByLabelText(/Branch name/i);
      });
    });

    describe('Detail attributes', () => {
      it('Should have a Detail attributes when Request Detail loaded.', async () => {
        screen.findByText(/Detail attributes/i);
      });

      it('Should have a Request template label in Detail', () => {
        screen.findByLabelText(/Request template/i);
      });

      it('Should have a Status label in Detail', () => {
        screen.findByLabelText(/Status/i);
      });

      it('Should have a Current state label in Detail', () => {
        screen.findByLabelText(/Current state/i);
      });

      it('Should have a Stakeholders label in Detail', () => {
        screen.findByLabelText(/Stakeholders/i);
      });

      it('Should have a Creation time label in Detail', () => {
        screen.findByLabelText(/Creation time/i);
      });
    });
  });
});
