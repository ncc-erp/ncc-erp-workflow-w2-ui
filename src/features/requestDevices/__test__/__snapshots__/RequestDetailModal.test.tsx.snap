// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Request Detail should match snapshot when rendering 1`] = `
<body>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    class="chakra-portal"
  >
    <div
      aria-hidden="true"
      class="chakra-modal__overlay css-13d1tkw"
      data-aria-hidden="true"
      style="opacity: 0;"
    />
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="chakra-modal__content-container css-1oxhx2p"
        tabindex="-1"
      >
        <section
          aria-describedby="chakra-modal--body-:r0:"
          aria-labelledby="chakra-modal--header-:r0:"
          aria-modal="true"
          class="chakra-modal__content css-ctar5y"
          id="chakra-modal-:r0:"
          role="dialog"
          style="opacity: 0; transform: scale(0.95) translateZ(0);"
          tabindex="-1"
        >
          <header
            class="chakra-modal__header css-72fd9l"
            id="chakra-modal--header-:r0:"
          >
            <div
              class="chakra-stack css-1igwmid"
            >
              <img
                class="chakra-image css-kow2w3"
                src=""
              />
              <h2
                class="chakra-heading css-sigo6f"
              >
                <p
                  class="chakra-text css-fb81nc"
                >
                  Office Equipment Request
                </p>
                <p
                  class="chakra-text css-issjhg"
                >
                  Request Details
                </p>
              </h2>
            </div>
            <button
              class="chakra-button css-12eh3jv"
              type="button"
            >
              View Workflow Detail
            </button>
          </header>
          <button
            aria-label="Close"
            class="chakra-modal__close-btn css-1m7215s"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                fill="currentColor"
              />
            </svg>
          </button>
          <div
            class="chakra-modal__body modalBody css-1qou8as"
            id="chakra-modal--body-:r0:"
          >
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-1urc9z9"
            />
            <div
              class="container"
            >
              <p
                class="chakra-text css-10g9qth"
              >
                Request input
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Current Office
                  </label>
                  <p
                    class="content "
                  >
                    <div>
                      Mocked MarkdownEditor
                    </div>
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Equipment
                  </label>
                  <p
                    class="content "
                  >
                    <div>
                      Mocked MarkdownEditor
                    </div>
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Reason
                  </label>
                  <p
                    class="content "
                  >
                    <div>
                      Mocked MarkdownEditor
                    </div>
                  </p>
                </div>
              </div>
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1urc9z9"
              />
              <p
                class="chakra-text css-10g9qth"
              >
                Request user
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Name
                  </label>
                  <p
                    class="content "
                  >
                    Nhan Huynh Ba
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Email
                  </label>
                  <p
                    class="content "
                  >
                    nhan.huynhba@ncc.asia
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Branch name
                  </label>
                  <p
                    class="content "
                  >
                    Đà Nẵng
                  </p>
                </div>
              </div>
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1urc9z9"
              />
              <p
                class="chakra-text css-yb4mb5"
              >
                Detail
              </p>
              <div
                class="wrapper"
              >
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Request template
                  </label>
                  <p
                    class="content "
                  >
                    Office Equipment Request
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Status
                  </label>
                  <p
                    class="content content-green"
                  >
                    Approved
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Current state
                  </label>
                  <p
                    class="content "
                  >
                    Branch Manager makes decision
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Stakeholders
                  </label>
                  <p
                    class="content "
                  >
                    Nhan Huynh Ba
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  >
                    Creation time
                  </label>
                  <p
                    class="content "
                  >
                    09-10-2023 11:49
                  </p>
                </div>
                <div
                  class="textGroup"
                >
                  <label
                    class="label"
                  />
                  <p
                    class="content "
                  />
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;
